openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -615,3 +617,216 @@ def test_user_change_with_no_ids(self):
+    def test_process_request_emails_match(
+        self, mock_set_custom_attribute, mock_mark_cookie_for_deletion
+    ):
+        """
+        Calls EmailChangeMiddleware.process_request when user is authenticated,
+        ENFORCE_SESSION_EMAIL_MATCH is enabled and user session and request email also match.
+        Verifies that session and cookies are not affected.
+        """
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.TEST_PASSWORD)
+        self.request.session = self.client.session
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # Registering email change (simulating email change requested)
+        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
+
+        # Ensure email is set in the session
+        self.assertEqual(self.request.session.get('email'), self.user.email)
+        # Ensure session cookie exist
+        self.assertEqual(len(self.client.response.cookies), 1)
+
+        # Call process_request
+        EmailChangeMiddleware(get_response=lambda request: None).process_request(self.request)
+
+        # Verify that session_email_match and is_enforce_session_email_match_enabled
+        # custom attributes are set
+        mock_set_custom_attribute.assert_has_calls([call('session_email_match', True)])
+        mock_set_custom_attribute.assert_has_calls([call('is_enforce_session_email_match_enabled', True)])
+
+        # Assert that the session and cookies are not affected
+        self.assertEqual(self.request.session.get('email'), self.user.email)
+        self.assertEqual(len(self.client.response.cookies), 1)
+        self.assertEqual(self.client.response.cookies[settings.SESSION_COOKIE_NAME].value, 'authenticated')
+
+        # Assert that _mark_cookie_for_deletion not called
+        mock_mark_cookie_for_deletion.assert_not_called()
+
+    @override_settings(ENFORCE_SESSION_EMAIL_MATCH=True)
+    @patch('openedx.core.djangoapps.safe_sessions.middleware._mark_cookie_for_deletion')
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_custom_attribute")
+    def test_process_request_emails_mismatch(
+        self, mock_set_custom_attribute, mock_mark_cookie_for_deletion
+    ):
+        """
+        Calls EmailChangeMiddleware.process_request when user is authenticated,
+        ENFORCE_SESSION_EMAIL_MATCH is enabled and user session and request email mismatch.
+        Verifies that session is flushed and cookies are marked for deletion.
+        """
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.TEST_PASSWORD)
+        self.request.session = self.client.session
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # Registering email change (simulating email change requested)
```

Comment:
1. First, does my suggestion make sense? Is this the difference between this call and the `# simulating email changed` below? Maybe that should rad
```suggestion
        # Simulates original email update (e.g. during account registration, or some past email change).
```
2. Are we missing a test for where there is no previous history of an email change (e.g. all existing sessions that predate this code)? We want to make sure we don't accidentally kill these sessions.

Reply:
> First, does my suggestion make sense? Is this the difference between this call and the # simulating email changed below? Maybe that should rad

At this point, when I think about the two purposes of the `register_email_change` function now. Do you think we need to have separate functions for both purposes?

Let me explain my point in detail. 

Purpose 1: We want to store user email in session on registration, login, and email change to reflect current email of user in session so that it can be used for comparison by the `process_request` function of middleware. 

Purpose 2: We want to set a variable in the request cache to indicate that the email was changed so we have to update logged-in cookies with the new email.

Purpose 2 is only required upon successful email change and NOT when the user registers or logs in.

(Although this is not a big deal) Do you think we should have them in separate functions? This will avoid creating and setting user's logged-in cookies.  (once when the login or registers and second on the process_response function of middleware). 
Also this may help us have more logical tests/comments? 

I would like your opinion on this? Maybe you have more appropriate suggestion on this?

Reply:
> Are we missing a test for where there is no previous history of an email change (e.g. all existing sessions that predate this code)? We want to make sure we don't accidentally kill these sessions.

I have added a test `test_process_request_no_email_change_history` for this purpose. Please have a look at that test.

Reply:
> Purpose 1: Stores user email in session on registration, login, and email change to reflect current email of user in session so that it can be used for comparison by the process_request function of middleware.

Should this just be "registration" and "email change"? It seems like "login" would typically not update anything, but instead should fail if mismatched, right? Maybe there is the exceptional case of "login" when session email has never been set before?

Note: I am writing these review comments without studying the code or seeing any of your latest changes. If these comments are not helpful, I plan to do a more complete review at some point.

> So I am not sure if we can say `Simulates original email update`? as it's not an email update here actually.

I meant something more like: `Simulates a past email change (e.g. from an earlier request)`.

Also, if instead of calling register the test first made a "registration" or (second test: "email change") before the request under test, it would ensure everything is hanging together appropriately in all cases. However, this might be the more complete integration test, and here we may just want the comment to clarify where this would have happened in a real situation.

> I have added a test `test_process_request_no_email_change_history` for this purpose. Please have a look at that test.

Thank you. I still need to get to the actual review.

Reply:
Marking this as resolved.